File: app/src/main/java/com/hussain/podcastapp/ui/EpisodesActivity.java
package com.hussain.podcastapp.ui;

import android.annotation.SuppressLint;
import android.app.ActivityOptions;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Html;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.hussain.podcastapp.R;
import com.hussain.podcastapp.adapter.EpisodeAdapter;
import com.hussain.podcastapp.base.BaseActivity;
import com.hussain.podcastapp.database.ApiInterface;
import com.hussain.podcastapp.model.Channel;
import com.hussain.podcastapp.model.Item;
import com.hussain.podcastapp.model.RssFeed;
import com.hussain.podcastapp.utils.AppConstants;
import com.hussain.podcastapp.utils.GlideApp;

import java.util.List;

import butterknife.BindView;
import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.simplexml.SimpleXmlConverterFactory;

public class EpisodesActivity extends BaseActivity implements EpisodeAdapter.EpisodeClickListener {

    private static final String TAG = EpisodesActivity.class.getName();

    @BindView(R.id.tvTitle)
    TextView mTitle;
    @BindView(R.id.tvDescription)
    TextView mDescription;
    @BindView(R.id.ivArtwork)
    ImageView mArtwork;
    @BindView(R.id.tvEpisodes)
    TextView mTvEpispodes;
    @BindView(R.id.rvEpisodes)
    RecyclerView rvEpisodes;
    @BindView(R.id.tvError)
    TextView mTvError;
    @BindView(R.id.error_layout)
    ConstraintLayout mErrorLayout;
    @Nullable
    @BindView(R.id.coordinatorLayout)
    CoordinatorLayout mEpisodeLayout;

    private String mArtworkUrl;
    private List<Item> mItems;
    private Channel mChannel;

    @SuppressLint("MissingSuperCall")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        onCreate(savedInstanceState, R.layout.activity_episodes);
        Intent b = getIntent();
        String mFeedUrl = b.getStringExtra(AppConstants.FEED_URL_KEY);
        mArtworkUrl = b.getStringExtra(AppConstants.ARTWORK_URL);
        if (mFeedUrl != null) {
            RssFeed(mFeedUrl);
            rvEpisodes.setLayoutManager(new LinearLayoutManager(this));
        }
    }

    private void setUI() {
        if (mChannel != null && mItems != null) {
            String episodesText = mItems.size() + " " + getResources().getString(R.string.episodes_text);
            mTitle.setText(mChannel.getTitle());
            mDescription.setText(Html.fromHtml(mChannel.getDescription()));
            mTvEpispodes.setText(episodesText);
            GlideApp.with(this)
                    .load(mArtworkUrl)
                    .placeholder(R.color.colorPrimary)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(mArtwork);
            EpisodeAdapter adapter = new EpisodeAdapter(mItems, this);
            rvEpisodes.setAdapter(adapter);
            showAnimation(false);
        }
    }

    private void RssFeed(String url) {
        showAnimation(true);
        Uri uri = Uri.parse(url);
        String protocol = uri.getScheme();
        String server = uri.getAuthority();
        String path = uri.getPath();
        String limit = uri.getQueryParameter("id");
        url = protocol + "://" + server + path + "/";
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(url)
                .client(new OkHttpClient())
                .addConverterFactory(SimpleXmlConverterFactory.create())
                .build();
        ApiInterface apiInterface = retrofit.create(ApiInterface.class);
        Call<RssFeed> rssFeedCall = apiInterface.getRssFeed(limit);
        rssFeedCall.enqueue(new Callback<RssFeed>() {
            @Override
            public void onResponse(@NonNull Call<RssFeed> call, @NonNull Response<RssFeed> response) {
                RssFeed data = response.body();
                if (data != null) {
                    mChannel = data.getChannel();
                    mItems = mChannel.getItems();
                    setUI();
                } else {
                    mEpisodeLayout.setVisibility(View.GONE);
                    mErrorLayout.setVisibility(View.VISIBLE);
                    mTvError.setText(R.string.no_podcast);
                    showAnimation(false);
                }
            }

            @Override
            public void onFailure(@NonNull Call<RssFeed> call, @NonNull Throwable t) {
                Log.d(TAG, t.getMessage());
                Toast.makeText(getApplicationContext(), R.string.user_error_text, Toast.LENGTH_LONG).show();
                finish();
            }
        });
    }

    @Override
    public void onEpisodeClick(Item item, int position, ImageView view) {
        Intent intent = new Intent(this, PlayerActivity.class);
        Bundle dataBundle = new Bundle();
        dataBundle.putParcelable(AppConstants.ITEM_KEY, item);
        dataBundle.putString(AppConstants.SHARE_KEY, mChannel.getShareLink());
        intent.putExtra(AppConstants.BUNDLE_KEY, dataBundle);
        Bundle transitionBundle = ActivityOptions.makeSceneTransitionAnimation(this, view, view.getTransitionName()).toBundle();
        startActivity(intent, transitionBundle);

    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putParcelable(AppConstants.SCROLL_POSITION, rvEpisodes.getLayoutManager().onSaveInstanceState());
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        rvEpisodes.getLayoutManager().onRestoreInstanceState(savedInstanceState.getParcelable(AppConstants.SCROLL_POSITION));
        super.onRestoreInstanceState(savedInstanceState);
    }

    @Override
    public void onToolBarSetUp(Toolbar toolbar, ActionBar actionBar) {
        TextView tvHeader = toolbar.findViewById(R.id.tvClassName);
        tvHeader.setText(R.string.app_name);
    }
}

File: app/src/main/java/com/hussain/podcastapp/ui/EpisodesActivity.kt
package com.hussain.podcastapp.ui

import android.annotation.SuppressLint
import android.app.ActivityOptions
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.text.Html
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.NonNull
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.InterstitialAd
import com.hussain.podcastapp.R
import com.hussain.podcastapp.adapter.EpisodeAdapter
import com.hussain.podcastapp.base.BaseActivity
import com.hussain.podcastapp.database.ApiInterface
import com.hussain.podcastapp.model.Channel
import com.hussain.podcastapp.model.Item
import com.hussain.podcastapp.model.RssFeed
import com.hussain.podcastapp.utils.AppConstants
import com.hussain.podcastapp.utils.GlideApp
import kotlinx.android.synthetic.main.activity_episodes.*
import kotlinx.android.synthetic.main.layout_error.*
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.simplexml.SimpleXmlConverterFactory

class EpisodesActivity : BaseActivity(), EpisodeAdapter.EpisodeClickListener {

    private var mInterstitialAd: InterstitialAd? = null
    private var mArtworkUrl: String? = null
    private var mItems: List<Item>? = null
    private var mChannel: Channel? = null

    @SuppressLint("MissingSuperCall")
    override fun onCreate(savedInstanceState: Bundle?) {
        onCreate(savedInstanceState, R.layout.activity_episodes)

        val b = intent
        val mFeedUrl = b.getStringExtra(AppConstants.FEED_URL_KEY)
        mArtworkUrl = b.getStringExtra(AppConstants.ARTWORK_URL)

        if (mFeedUrl != null) {
            getRssFeed(mFeedUrl)
            rvEpisodes?.layoutManager = LinearLayoutManager(this)
        }

        mInterstitialAd = InterstitialAd(this)
        mInterstitialAd?.adUnitId = "ca-app-pub-4616631912723447/5671156240"
        mInterstitialAd?.loadAd(AdRequest.Builder().build())
    }

    private fun getRssFeed(url: String) {
        var rssURL = url
        showAnimation(true)
        val uri = Uri.parse(rssURL)
        val protocol = uri.scheme
        val server = uri.authority
        val path = uri.path
        val limit = uri.getQueryParameter("id")
        rssURL = "$protocol://$server$path/"

        val retrofit = Retrofit.Builder()
                .baseUrl(rssURL)
                .client(OkHttpClient())
                .addConverterFactory(SimpleXmlConverterFactory.create())
                .build()

        val apiInterface = retrofit.create(ApiInterface::class.java)
        val rssFeedCall = apiInterface.getRssFeed(limit!!)

        rssFeedCall.enqueue(object : Callback<RssFeed> {
            override fun onResponse(@NonNull call: Call<RssFeed>, @NonNull response: Response<RssFeed>) {
                val data = response.body()
                if (data != null) {
                    mChannel = data.channel
                    mItems = mChannel!!.items
                    setUI()
                } else {
                    coordinatorLayout.visibility = View.GONE
                    error_layout.visibility = View.VISIBLE
                    tvError.setText(R.string.no_podcast)
                    showAnimation(false)
                }
            }

            override fun onFailure(@NonNull call: Call<RssFeed>, @NonNull t: Throwable) {
                Log.d(TAG, t.message)
                Toast.makeText(applicationContext, R.string.user_error_text, Toast.LENGTH_LONG).show()
                finish()
            }
        })
    }

    private fun setUI() {
        if (mChannel != null && mItems != null) {
            val episodesText = mItems?.size.toString() + " " + resources.getString(R.string.episodes_text)
            tvTitle.text = mChannel?.title
            tvDescription.text = Html.fromHtml(mChannel?.description)
            tvEpisodes.text = episodesText
            GlideApp.with(this)
                    .load(mArtworkUrl)
                    .placeholder(R.color.colorPrimary)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(ivArtwork)
            val adapter = EpisodeAdapter(mItems!!, this)
            rvEpisodes?.adapter = adapter
            showAnimation(false)
        }
        if (mInterstitialAd!!.isLoaded) {
            mInterstitialAd?.show()
        }
    }

    override fun onEpisodeClick(item: Item, position: Int, imageView: ImageView?) {
        val intent = Intent(this, PlayerActivity::class.java)
        val dataBundle = Bundle()
        dataBundle.putParcelable(AppConstants.ITEM_KEY, item)
        dataBundle.putString(AppConstants.SHARE_KEY, mChannel?.shareLink)
        intent.putExtra(AppConstants.BUNDLE_KEY, dataBundle)
        val transitionBundle = ActivityOptions.makeSceneTransitionAnimation(this, imageView, imageView!!.transitionName).toBundle()
        startActivity(intent, transitionBundle)

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    public override fun onSaveInstanceState(@NonNull outState: Bundle) {
        outState.putParcelable(AppConstants.SCROLL_POSITION, rvEpisodes?.layoutManager?.onSaveInstanceState())
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        rvEpisodes?.layoutManager?.onRestoreInstanceState(savedInstanceState.getParcelable<Parcelable>(AppConstants.SCROLL_POSITION))
        super.onRestoreInstanceState(savedInstanceState)
    }

    override fun onToolBarSetUp(toolbar: Toolbar?, actionBar: ActionBar) {
        val tvHeader = toolbar!!.findViewById<TextView>(R.id.tvClassName)
        tvHeader.setText(R.string.app_name)
    }

    companion object {

        private val TAG = EpisodesActivity::class.java.name
    }
}

